crates/matrix-sdk-crypto/src/olm/group_sessions/sender_data_finder.rs

```diff
@@ -0,0 +1,891 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::ops::Deref;
+
+use async_trait::async_trait;
+use ruma::UserId;
+use tracing::error;
+use vodozemac::Curve25519PublicKey;
+
+use super::{SenderData, SenderDataRetryDetails};
+use crate::{
+    error::OlmResult,
+    identities::Device,
+    store::{self, Store},
+    types::{events::olm_v1::DecryptedRoomKeyEvent, DeviceKeys, MasterPubkey},
+    EventError, OlmError, OlmMachine, ReadOnlyDevice, ReadOnlyOwnUserIdentity,
+    ReadOnlyUserIdentities,
+};
+
+/// Temporary struct that is used to look up [`SenderData`] based on the
+/// information supplied with in [`InboundGroupSession`].
+///
+/// The letters A, B etc. in the documentation refer to the algorithm described
+/// in https://github.com/matrix-org/matrix-rust-sdk/issues/3543
+pub(crate) struct SenderDataFinder<'a, S: FinderCryptoStore> {
+    own_crypto_store: &'a S,
+    own_user_id: &'a UserId,
+}
+
+impl<'a> SenderDataFinder<'a, Store> {
+    /// As described in https://github.com/matrix-org/matrix-rust-sdk/issues/3543
+    /// and https://github.com/matrix-org/matrix-rust-sdk/issues/3544
+    /// find the device info associated with the to-device message used to
+    /// create the InboundGroupSession we are about to create, and decide
+    /// whether we trust the sender.
+    pub(crate) async fn find_using_event(
+        own_olm_machine: &'a OlmMachine,
+        sender_curve_key: Curve25519PublicKey,
+        room_key_event: &'a DecryptedRoomKeyEvent,
+    ) -> OlmResult<SenderData> {
+        let finder = Self {
+            own_crypto_store: own_olm_machine.store(),
+            own_user_id: own_olm_machine.user_id(),
+        };
+        finder.have_event(sender_curve_key, room_key_event).await
+    }
+
+    /// As described in https://github.com/matrix-org/matrix-rust-sdk/issues/3543
+    /// and https://github.com/matrix-org/matrix-rust-sdk/issues/3544
+    /// use the supplied device info to decide whether we trust the sender.
+    pub(crate) async fn find_using_device_keys(
+        own_olm_machine: &'a OlmMachine,
+        device_keys: DeviceKeys,
+    ) -> OlmResult<SenderData> {
+        let finder = Self {
+            own_crypto_store: own_olm_machine.store(),
+            own_user_id: own_olm_machine.user_id(),
+        };
+        finder.have_device_keys(&device_keys).await
+    }
+}
+
+impl<'a, S: FinderCryptoStore> SenderDataFinder<'a, S> {
+    async fn have_event(
+        &self,
+        sender_curve_key: Curve25519PublicKey,
+        room_key_event: &'a DecryptedRoomKeyEvent,
+    ) -> OlmResult<SenderData> {
+        // A (start)
+        //
+        // TODO: take the session lock for session_id
```

Comment:
The `InboundGroupSession`. We expect multiple tasks to be possibly updating a session at the same time e.g. an async task spawned from one of the steps below and a retry from the background task that will come in a later PR.

Reply:
This conversation can wait for the next PR, which will be quite different because of the conversation in the other comment: https://github.com/matrix-org/matrix-rust-sdk/pull/3590#discussion_r1671860791